whistl/classifier.py

```diff
@@ -3,106 +3,212 @@
 import argparse
 import logging
 import numpy as np
-import os
-import pickle
 import random
+import time
 
 import torch
+from torch.autograd import grad
 import torch.nn as nn
 import torch.optim as optim
 from torch.utils.data import DataLoader
+from tqdm import tqdm
 
 import dataset
 import model
 import util
 
 
-def count_correct(output, labels):
-    '''Calculate the number of correct predictions in the given batch'''
-    # This could be more efficient with a hard sigmoid or something,
-    # Performance impact should be negligible though
-    correct = 0
-    predictions = [1 if p > .5 else 0 for p in output]
-    for y, y_hat in zip(predictions, labels):
-        if y == y_hat:
-            correct += 1
-    return correct
+def compute_irm_penalty(loss, dummy_w):
+    ''' Calculate the invariance penalty for the classifier'''
+    dummy_grad = abs(grad(loss, dummy_w, create_graph=True)[0])
 
+    return dummy_grad
 
-def parse_map_file(map_file_path):
-    '''Create a sample: label mapping from the pickled file output by label_samples.py
+
+def train_with_irm(classifier, map_file, train_dirs, tune_dirs, gene_file,
+                   num_epochs, loss_scaling_factor, label_to_encoding, device, logger=None):
+    '''Train the provided classifier using invariant risk minimization
+    IRM looks for features in the data that are invariant between different environments, as
+    they are more likely to be predictive of true causal signals as opposed to spurious
+    correlations. For more information, read https://arxiv.org/abs/1907.02893
```

Comment:
General conceptual question: what do you expect IRM to give you in the context of your project? Do you think it will just give you better predictions, or does it have an interpretability advantage as well (i.e. some sort of measurement of which features are invariant/possibly causal)? If the latter, is there a simple way to see which features are invariant?

Reply:
The authors claim that the data representation you get from training a model with IRM is more able to learn which features truly matter, by training your model to perform well in different environments. The questions I'm currently asking are
1) Does IRM work with a large number of environments (like the number of studies related to a disease
2) Does IRM give a generalization boost by helping the model ignore batch effects